crates/re_viewer/src/misc/tensor_to_gpu.rs

```diff
@@ -0,0 +1,398 @@
+        (3, TensorData::U16(buf)) => (pad_and_cast(buf, 0), TextureFormat::Rgba16Uint),
+        (3, TensorData::I16(buf)) => (pad_and_cast(buf, 0), TextureFormat::Rgba16Sint),
+        (3, TensorData::U32(buf)) => (pad_and_cast(buf, 0), TextureFormat::Rgba32Uint),
+        (3, TensorData::I32(buf)) => (pad_and_cast(buf, 0), TextureFormat::Rgba32Sint),
+        // (3, TensorData::F16(buf)) => (pad_and_cast(buf, 0.0), TextureFormat::Rgba16Float), TODO(#854)
+        (3, TensorData::F32(buf)) => (pad_and_cast(buf, 0.0), TextureFormat::Rgba32Float),
+        (3, TensorData::F64(buf)) => {
+            let pad = 0.0;
+            let floats: Vec<f32> = buf
+                .chunks_exact(3)
+                .flat_map(|chunk| [chunk[0] as f32, chunk[1] as f32, chunk[2] as f32, pad])
+                .collect();
+            (
+                pod_collect_to_vec(&floats).into(),
+                TextureFormat::Rgba32Float,
+            )
+        }
+
+        // TODO(emilk): premultiply alpha
+        (4, TensorData::U8(buf)) => (cast_slice_to_cow(buf.as_slice()), TextureFormat::Rgba8Uint),
+        (4, TensorData::I8(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba8Sint),
+        (4, TensorData::U16(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba16Uint),
+        (4, TensorData::I16(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba16Sint),
+        (4, TensorData::U32(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba32Uint),
+        (4, TensorData::I32(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba32Sint),
+        // (4, TensorData::F16(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba16Float), TODO(#854)
+        (4, TensorData::F32(buf)) => (cast_slice_to_cow(buf), TextureFormat::Rgba32Float),
+        (4, TensorData::F64(buf)) => (narrow_f64_to_f32s(buf), TextureFormat::Rgba32Float),
+
+        // TODO(emilk): U64/I64
+        (_depth, dtype) => {
+            anyhow::bail!("Don't know how to turn a tensor of shape={:?} and dtype={dtype:?} into a color image", tensor.shape)
+        }
+    };
+
+    Ok(Texture2DCreationDesc {
+        label: debug_name.into(),
+        data,
+        format,
+        width,
+        height,
+    })
+}
+
+fn get_or_create_texture<'a, Err>(
+    render_ctx: &mut RenderContext,
+    texture_key: u64,
+    try_create_texture_desc: impl FnOnce() -> Result<Texture2DCreationDesc<'a>, Err>,
+) -> Result<GpuTexture2DHandle, Err> {
+    render_ctx.texture_manager_2d.get_or_create_with(
+        texture_key,
+        &mut render_ctx.gpu_resources.textures,
+        try_create_texture_desc,
+    )
+}
+
+fn cast_slice_to_cow<From: Pod>(slice: &[From]) -> Cow<'_, [u8]> {
+    cast_slice(slice).into()
+}
+
+// wgpu doesn't support f64 textures, so we need to narrow to f32:
+fn narrow_f64_to_f32s(slice: &[f64]) -> Cow<'static, [u8]> {
+    crate::profile_function!();
+    let f32s: Vec<f32> = slice.iter().map(|&f| f as f32).collect::<Vec<f32>>();
+    let bytes: Vec<u8> = pod_collect_to_vec(&f32s);
```

Comment:
as far as I can tell all uses of `pod_collect_to_vec` here should either actually do the collecting instead of taking a vec, or, use https://docs.rs/bytemuck/latest/bytemuck/allocation/fn.cast_vec.html to cast the vec directly without allocating new memory/memcpying